<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Genome-wide analysis — cursus â€¢ globalSeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Google analytics -->
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-96845398-2', 'auto');
  ga('send', 'pageview');

</script>

  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">globalSeq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">functions</a>
</li>
<li>
  <a href="../articles/vignette.html">vignette</a>
</li>
<li>
  <a href="../articles/article.html">article</a>
</li>
<li>
  <a href="../news/index.html">news</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rauschenberger/globalSeq">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://bioconductor.org/packages/globalSeq/">
    <span class="fa fa-download"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Genome-wide analysis</h1>
    </div>

    
    <p>This function tests for associations between gene expression
or exon abundance (<code>Y</code>)
and genetic or epigenetic alterations (<code>X</code>).
Using the locations of genes (<code>Yloc</code>),
and the locations of genetic
or epigenetic alterations (<code>Xloc</code>),
the expression of each gene is tested for associations with 
alterations on the same chromosome that are closer to the gene
than a given distance (<code>window</code>).</p>
    

    <pre class="usage"><span class='fu'>cursus</span>(<span class='no'>Y</span>, <span class='no'>Yloc</span>, <span class='no'>X</span>, <span class='no'>Xloc</span>, <span class='no'>window</span>,
        <span class='kw'>Ychr</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>Xchr</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
        <span class='kw'>offset</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>group</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
        <span class='kw'>perm</span> <span class='kw'>=</span> <span class='fl'>1000</span>, <span class='kw'>nodes</span> <span class='kw'>=</span> <span class='fl'>2</span>,
        <span class='kw'>phi</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>kind</span> <span class='kw'>=</span> <span class='fl'>0.01</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>Y</th>
      <td><p><strong>RNA-Seq data</strong><strong>:</strong>
numeric matrix with <code>q</code> rows (genes)
and <code>n</code> columns (samples);
or a SummarizedExperiment object</p></td>
    </tr>
    <tr>
      <th>Yloc</th>
      <td><p><strong>location RNA-Seq</strong><strong>:</strong>
numeric vector of length <code>q</code>
(point location)<strong>;</strong>
numeric matrix with <code>q</code> rows
and two columns (start and end locations)</p></td>
    </tr>
    <tr>
      <th>X</th>
      <td><p><strong>genomic profile</strong><strong>:</strong>
numeric matrix with <code>p</code> rows (covariates)
and <code>n</code> columns (samples)</p></td>
    </tr>
    <tr>
      <th>Xloc</th>
      <td><p><strong>location covariates</strong><strong>:</strong>
numeric vector of length <code>p</code></p></td>
    </tr>
    <tr>
      <th>window</th>
      <td><p><strong>maximum distance</strong><strong>:</strong>
non-negative real number</p></td>
    </tr>
    <tr>
      <th>Ychr</th>
      <td><p>chromosome RNA-Seq<strong>:</strong>
factor of length <code>q</code></p></td>
    </tr>
    <tr>
      <th>Xchr</th>
      <td><p>chromosome covariates<strong>:</strong>
factor of length <code>p</code></p></td>
    </tr>
    <tr>
      <th>offset</th>
      <td><p>numeric vector of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>group</th>
      <td><p>confounding variable<strong>:</strong>
factor of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>perm</th>
      <td><p>number of iterations<strong>:</strong>
positive integer</p></td>
    </tr>
    <tr>
      <th>nodes</th>
      <td><p>number of cluster nodes for parallel computation</p></td>
    </tr>
    <tr>
      <th>phi</th>
      <td><p>dispersion parameters<strong>:</strong> vector of length <code>q</code></p></td>
    </tr>
    <tr>
      <th>kind</th>
      <td><p>computation <strong>:</strong>
number between 0 and 1</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The function returns a dataframe,
with the p-values in the first row
and the test statistics in the second row.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Note that <code>Yloc</code>, <code>Xloc</code> and <code>window</code> must
be given in the same unit, usually in base pairs.
If <code>Yloc</code> indicates interval <strong>locations</strong>,
and <code>window</code> is zero,
then only covariates between the start and end location
of the gene are of interest.
Typically <code>window</code> is larger than one million base pairs.</p>
<p>If <code>Y</code> and <code>X</code> include data from a single chromosome,
<code>Ychr</code> and <code>Xchr</code> are redundant.
If <code>Y</code> or <code>X</code> include data
from <strong>multiple chromosomes</strong>,
<code>Ychr</code> and <code>Xchr</code> should be specified
in order to prevent confusion between chromosomes.</p>
<p>For the simultaneous analysis of
<strong>multiple genomic profiles</strong>
<code>X</code> should be a list of numeric matrices with
<code>n</code> columns (samples),
<code>Xloc</code> a list of numeric vectors,
and <code>window</code> a list of non-negative real numbers.
If provided, <code>Xchr</code> should be alist of of numeric vectors.</p>
<p>The <code>offset</code> is meant to account for
different <strong>libary sizes</strong>.
By default the <code>offset</code> is calculated based on <code>Y</code>.
Different library sizes can be ignored by 
setting the <code>offset</code> to <code>rep(1,n)</code>.</p>
<p>The user can provide the <strong>confounding</strong> variable <code>group</code>.
Note that each level of <code>group</code> must appear at least twice
in order to allow stratified permutations.</p>
<p>Efficient alternatives to classical <strong>permutation</strong> (<code>kind=1</code>)
are the method of control variates (<code>kind=0</code>)
and permutation in chunks (0 &lt; <code>kind</code> &lt; 1)
<a href='intern.crude.html'>details</a>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>A Rauschenberger, MA Jonker, MA van de Wiel, and RX Menezes (2016).
"Testing for association between RNA-Seq and high-dimensional data",
<em>BMC Bioinformatics</em>. 17:118.
<a href='http://dx.doi.org/10.1186/s12859-016-0961-5'>html</a>
<a href='http://www.biomedcentral.com/content/pdf/s12859-016-0961-5.pdf'>pdf</a>
(open access)</p>
<p>RX Menezes, M Boetzer, M Sieswerda, GJB van Ommen, and JM Boer (2009).
"Integrated analysis of DNA copy number
and gene expression microarray data using gene sets",
<em>BMC Bioinformatics</em>. 10:203.
<a href='http://dx.doi.org/10.1186/1471-2105-10-203'>html</a>
<a href='http://www.biomedcentral.com/content/pdf/1471-2105-10-203.pdf'>pdf</a>
(open access)</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p>The function <code><a href='omnibus.html'>omnibus</a></code> tests for associations
between an overdispersed response variable
and a high-dimensional covariate set.
The function <code><a href='proprius.html'>proprius</a></code> calculates the contributions
of individual samples or covariates to the test statistic.
All other function of the R package
<code><a href='globalSeq-package.html'>globalSeq</a></code> are <code><a href='internal.html'>internal</a></code>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># simulate high-dimensional data</span>
<span class='no'>n</span> <span class='kw'>&lt;-</span> <span class='fl'>30</span>; <span class='no'>q</span> <span class='kw'>&lt;-</span> <span class='fl'>10</span>; <span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fl'>100</span>
<span class='no'>Y</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnbinom</span>(<span class='no'>q</span>*<span class='no'>n</span>,<span class='kw'>mu</span><span class='kw'>=</span><span class='fl'>10</span>,
    <span class='kw'>size</span><span class='kw'>=</span><span class='fl'>1</span>/<span class='fl'>0.25</span>),<span class='kw'>nrow</span><span class='kw'>=</span><span class='no'>q</span>,<span class='kw'>ncol</span><span class='kw'>=</span><span class='no'>n</span>)
<span class='no'>X</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnorm</span>(<span class='no'>p</span>*<span class='no'>n</span>),<span class='kw'>nrow</span><span class='kw'>=</span><span class='no'>p</span>,<span class='kw'>ncol</span><span class='kw'>=</span><span class='no'>n</span>)
<span class='no'>Yloc</span> <span class='kw'>&lt;-</span> <span class='fu'>seq</span>(<span class='fl'>0</span>,<span class='fl'>1</span>,<span class='kw'>length.out</span><span class='kw'>=</span><span class='no'>q</span>)
<span class='no'>Xloc</span> <span class='kw'>&lt;-</span> <span class='fu'>seq</span>(<span class='fl'>0</span>,<span class='fl'>1</span>,<span class='kw'>length.out</span><span class='kw'>=</span><span class='no'>p</span>)
<span class='no'>window</span> <span class='kw'>&lt;-</span> <span class='fl'>1</span>

<span class='co'># hypothesis testing</span>
<span class='fu'>cursus</span>(<span class='no'>Y</span>,<span class='no'>Yloc</span>,<span class='no'>X</span>,<span class='no'>Xloc</span>,<span class='no'>window</span>)</div><div class='output co'>#&gt; <span class='message'>Analysing a single chromosome.</span></div><div class='output co'>#&gt;    pvalue   teststat covs
#&gt; 1   0.009  12.608557  100
#&gt; 2   1.000   6.955576  100
#&gt; 3   1.000  -5.856714  100
#&gt; 4   1.000  -4.170157  100
#&gt; 5   1.000  -2.679116  100
#&gt; 6   1.000  -2.845209  100
#&gt; 7   1.000  -1.478433  100
#&gt; 8   1.000  -5.910395  100
#&gt; 9   1.000 -11.607122  100
#&gt; 10  1.000  -6.922350  100</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Armin Rauschenberger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
