<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Internal function — intern.crude • globalSeq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Internal function — intern.crude" />

<meta property="og:description" content="Using the parameter estimates mu and phi
and the permutation matrix perm, these functions
tests for global association between y and X.
The function intern.crude calculates
p-values by permutation (without repetitions).
The functions intern.focus and
intern.conva use different tricks
to increase precision and decrease computational expense." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">globalSeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.11.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">functions</a>
</li>
<li>
  <a href="../articles/vignette.html">vignette</a>
</li>
<li>
  <a href="../articles/article.html">article</a>
</li>
<li>
  <a href="../news/index.html">news</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rauschenberger/globalSeq">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://bioconductor.org/packages/globalSeq/">
    <span class="fa fa-download"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Internal function</h1>
    <small class="dont-index">Source: <a href='https://github.com/rauschenberger/globalSeq/blob/master/R/internal_functions.R'><code>R/internal_functions.R</code></a></small>
    <div class="hidden name"><code>intern.crude.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Using the parameter estimates <code>mu</code> and <code>phi</code>
and the permutation matrix <code>perm</code>, these functions
tests for global association between <code>y</code> and <code>X</code>.
The function <code>intern.crude</code> calculates
p-values by permutation (without repetitions).
The functions <code>intern.focus</code> and
<code>intern.conva</code> use different tricks
to increase precision and decrease computational expense.</p>
    
    </div>

    <pre class="usage"><span class='fu'>intern.crude</span>(<span class='no'>y</span>, <span class='no'>X</span>, <span class='no'>mu</span>, <span class='no'>phi</span>, <span class='no'>perm</span>)

<span class='fu'>intern.focus</span>(<span class='no'>y</span>, <span class='no'>X</span>, <span class='no'>mu</span>, <span class='no'>phi</span>, <span class='no'>perm</span>, <span class='no'>focus</span>)

<span class='fu'>intern.conva</span>(<span class='no'>y</span>, <span class='no'>X</span>, <span class='no'>mu</span>, <span class='no'>phi</span>, <span class='no'>perm</span>, <span class='no'>offset</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>y</th>
      <td><p>response variable: numeric vector of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>X</th>
      <td><p>covariate set: numeric matrix with <code>n</code> rows (samples)
and <code>p</code> columns (covariates)</p></td>
    </tr>
    <tr>
      <th>mu</th>
      <td><p>mean parameters: numeric vector of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>phi</th>
      <td><p>dispersion parameter: non-negative real number</p></td>
    </tr>
    <tr>
      <th>perm</th>
      <td><p>permutations: matrix with <code>n</code> rows (see example)</p></td>
    </tr>
    <tr>
      <th>focus</th>
      <td><p>number between 0 and 1</p></td>
    </tr>
    <tr>
      <th>offset</th>
      <td><p>numeric vector of length <code>n</code></p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Each function returns a dataframe,
with the p-value in the first row,
and the test statistic in the second row.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The function <code>intern.focus</code>
uses permutations in chunks.
If the remaining permutations do not allow
to reach a specified significance level,
it stops and rounds the p-value to one.</p>
<p>The function <code>intern.conva</code>
uses the method of control variates
from Senchaudhuri et al. (1995).
Roughly speaking, if the test statistics
from Rauschenberger et al. (2016)
and Goeman et al. (2004) are highly correlated,
it returns the asymptotic p-value from Goeman et al. (2004).</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>P Senchaudhuri, CR Mehta, and NR Patel (1995).
"Estimating exact p values by the method of control variates
or Monte Carlo rescue",
<em>Journal of the American Statistical Association</em>.
90:640-648
<a href='http://dx.doi.org/10.2307/2291077'>html</a>
<a href='http://www.jstor.org/stable/pdf/2291077.pdf?acceptTC=true'>pdf</a>
(restricted access)</p>
<p>A Rauschenberger, MA Jonker, MA van de Wiel, and RX Menezes (2016).
"Testing for association between RNA-Seq and high-dimensional data",
<em>BMC Bioinformatics</em>. 17:118.
<a href='http://dx.doi.org/10.1186/s12859-016-0961-5'>html</a>
<a href='http://www.biomedcentral.com/content/pdf/s12859-016-0961-5.pdf'>pdf</a>
(open access)</p>
<p>JJ Goeman, SA van de Geer, F de Kort, and HC van Houwelingen (2004).
"A global test for groups of genes:
testing association with a clinical outcome",
<em>Bioinformatics</em>. 20:93-99.
<a href='http://dx.doi.org/10.1093/bioinformatics/btg382'>html</a>
<a href='http://bioinformatics.oxfordjournals.org/content/20/1/93.full.pdf'>pdf</a>
(open access)</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>These are <code><a href='internal.html'>internal</a></code> functions. The user functions 
of the R package <code><a href='globalSeq-package.html'>globalSeq</a></code> are <code><a href='cursus.html'>cursus</a></code>,
<code><a href='omnibus.html'>omnibus</a></code>, and <code><a href='proprius.html'>proprius</a></code>.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># simulate high-dimensional data</span>
<span class='no'>n</span> <span class='kw'>&lt;-</span> <span class='fl'>30</span>
<span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fl'>100</span>
<span class='co'># set.seed(1)</span>
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/NegBinomial'>rnbinom</a></span>(<span class='no'>n</span>,<span class='kw'>mu</span><span class='kw'>=</span><span class='fl'>10</span>,<span class='kw'>size</span><span class='kw'>=</span><span class='fl'>1</span>/<span class='fl'>0.25</span>)
<span class='no'>X</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/matrix'>matrix</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>(<span class='no'>n</span>*<span class='no'>p</span>),<span class='kw'>nrow</span><span class='kw'>=</span><span class='no'>n</span>,<span class='kw'>ncol</span><span class='kw'>=</span><span class='no'>p</span>)

<span class='co'># prepare arguments</span>
<span class='no'>mu</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/rep'>rep</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/mean'>mean</a></span>(<span class='no'>y</span>),<span class='no'>n</span>)
<span class='no'>phi</span> <span class='kw'>&lt;-</span> (<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/cor'>var</a></span>(<span class='no'>y</span>)-<span class='no'>mu</span>)/<span class='no'>mu</span>^<span class='fl'>2</span>
<span class='no'>perm</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='intern.permu.html'>intern.permu</a></span>(<span class='kw'>n</span><span class='kw'>=</span><span class='no'>n</span>,<span class='kw'>it</span><span class='kw'>=</span><span class='fl'>99</span>,<span class='kw'>group</span><span class='kw'>=</span><span class='kw'>NULL</span>,<span class='kw'>kind</span><span class='kw'>=</span><span class='fl'>1</span>)

<span class='co'># perform tests</span>
<span class='fu'>intern.crude</span>(<span class='no'>y</span>,<span class='no'>X</span>,<span class='no'>mu</span>,<span class='no'>phi</span>,<span class='no'>perm</span>)</div><div class='output co'>#&gt;   pvalue teststat covs
#&gt; 1   0.13  3.41357  100</div><div class='input'><span class='fu'>intern.focus</span>(<span class='no'>y</span>,<span class='no'>X</span>,<span class='no'>mu</span>,<span class='no'>phi</span>,<span class='no'>perm</span>,<span class='kw'>focus</span><span class='kw'>=</span><span class='fl'>0.01</span>)</div><div class='output co'>#&gt;   pvalue teststat covs
#&gt; 1      1  3.41357  100</div><div class='input'><span class='fu'>intern.conva</span>(<span class='no'>y</span>,<span class='no'>X</span>,<span class='no'>mu</span>,<span class='no'>phi</span>,<span class='no'>perm</span>,<span class='kw'>NULL</span>)</div><div class='output co'>#&gt;      pvalue teststat covs    rausch     goeman  cor     pstar
#&gt; 1 0.1634231  3.41357  100 0.1212121 0.09090909 0.85 0.1347709</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://rauschenberger.github.io'>Armin Rauschenberger</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

