<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Internal function — intern.chromo • globalSeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Internal function — intern.chromo" />

<meta property="og:description" content="Communicates between cursus and omnibus
by coordinating a chromosome-wide analysis." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">globalSeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">functions</a>
</li>
<li>
  <a href="../articles/vignette.html">vignette</a>
</li>
<li>
  <a href="../articles/article.html">article</a>
</li>
<li>
  <a href="../news/index.html">news</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rauschenberger/globalSeq">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://bioconductor.org/packages/globalSeq/">
    <span class="fa fa-download"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Internal function</h1>
    <small class="dont-index">Source: <a href='https://github.com/rauschenberger/globalSeq/blob/master/R/internal_functions.R'><code>R/internal_functions.R</code></a></small>
    <div class="hidden name"><code>intern.chromo.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Communicates between <code><a href='cursus.html'>cursus</a></code> and <code><a href='omnibus.html'>omnibus</a></code>
by coordinating a chromosome-wide analysis.</p>
    
    </div>

    <pre class="usage"><span class='fu'>intern.chromo</span>(<span class='no'>Y</span>, <span class='no'>Ystart</span>, <span class='no'>Yend</span>, <span class='no'>X</span>, <span class='no'>Xloc</span>,
             <span class='no'>window</span>, <span class='no'>offset</span>, <span class='no'>group</span>, <span class='no'>perm</span>,
             <span class='no'>nodes</span>, <span class='no'>phi</span>, <span class='no'>kind</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>Y</th>
      <td><p>RNA-Seq data<strong>:</strong>
numeric matrix with <code>q</code> rows (genes)
and <code>n</code> columns (samples);
or a SummarizedExperiment object</p></td>
    </tr>
    <tr>
      <th>Ystart</th>
      <td><p>start location of genes<strong>:</strong>
numeric vector of length <code>q</code></p></td>
    </tr>
    <tr>
      <th>Yend</th>
      <td><p>end location of genes<strong>:</strong>
NULL or numeric vector of length <code>q</code></p></td>
    </tr>
    <tr>
      <th>X</th>
      <td><p>genomic profile<strong>:</strong>
numeric matrix with <code>p</code> rows (covariates)
and <code>n</code> columns (samples)</p></td>
    </tr>
    <tr>
      <th>Xloc</th>
      <td><p>location covariates<strong>:</strong>
numeric vector of length <code>p</code></p></td>
    </tr>
    <tr>
      <th>window</th>
      <td><p>maximum distance<strong>:</strong>
non-negative real number</p></td>
    </tr>
    <tr>
      <th>offset</th>
      <td><p>numeric vector of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>group</th>
      <td><p>confounding variable<strong>:</strong>
factor of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>perm</th>
      <td><p>number of iterations<strong>:</strong>
positive integer</p></td>
    </tr>
    <tr>
      <th>nodes</th>
      <td><p>number of cluster nodes for parallel computation</p></td>
    </tr>
    <tr>
      <th>phi</th>
      <td><p>dispersion parameters<strong>:</strong> vector of length <code>q</code></p></td>
    </tr>
    <tr>
      <th>kind</th>
      <td><p>computation <strong>:</strong>
number between 0 and 1</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The function returns a dataframe,
with the p-value in the first column,
and the test statistic in the second column.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># simulate high-dimensional data</span>
<span class='no'>n</span> <span class='kw'>&lt;-</span> <span class='fl'>30</span>
<span class='no'>q</span> <span class='kw'>&lt;-</span> <span class='fl'>10</span>
<span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fl'>100</span>
<span class='fu'>set.seed</span>(<span class='fl'>1</span>)
<span class='no'>Y</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnbinom</span>(<span class='no'>q</span>*<span class='no'>n</span>,<span class='kw'>mu</span><span class='kw'>=</span><span class='fl'>10</span>,
    <span class='kw'>size</span><span class='kw'>=</span><span class='fl'>1</span>/<span class='fl'>0.25</span>),<span class='kw'>nrow</span><span class='kw'>=</span><span class='no'>q</span>,<span class='kw'>ncol</span><span class='kw'>=</span><span class='no'>n</span>)
<span class='no'>X</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnorm</span>(<span class='no'>p</span>*<span class='no'>n</span>),<span class='kw'>nrow</span><span class='kw'>=</span><span class='no'>p</span>,<span class='kw'>ncol</span><span class='kw'>=</span><span class='no'>n</span>)
<span class='no'>Yloc</span> <span class='kw'>&lt;-</span> <span class='fu'>seq</span>(<span class='fl'>0</span>,<span class='fl'>1</span>,<span class='kw'>length.out</span><span class='kw'>=</span><span class='no'>q</span>)
<span class='no'>Xloc</span> <span class='kw'>&lt;-</span> <span class='fu'>seq</span>(<span class='fl'>0</span>,<span class='fl'>1</span>,<span class='kw'>length.out</span><span class='kw'>=</span><span class='no'>p</span>)
<span class='no'>window</span> <span class='kw'>&lt;-</span> <span class='fl'>1</span>

<span class='co'># hypothesis testing</span>
<span class='fu'><a href='cursus.html'>cursus</a></span>(<span class='no'>Y</span>,<span class='no'>Yloc</span>,<span class='no'>X</span>,<span class='no'>Xloc</span>,<span class='no'>window</span>)</div><div class='output co'>#&gt; <span class='message'>Analysing a single chromosome.</span></div><div class='output co'>#&gt; <span class='error'>Error in checkForRemoteErrors(val): 2 nodes produced errors; first error: there is no package called 'globalSeq'</span></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Armin Rauschenberger.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

