<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Omnibus test — omnibus • globalSeq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Omnibus test — omnibus" />

<meta property="og:description" content="Test of association between a count response and
one or more covariate sets.
This test may be conceptualised as
a test of overall significance in regression analysis,
where the response variable is overdispersed, and where
the number of explanatory variables (p)
exceeds the sample size (n).
The negative binomial distribution accounts for overdispersion
and a random effect model accounts for high dimensionality
(p&gt;&gt;n)." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">globalSeq</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">functions</a>
</li>
<li>
  <a href="../articles/vignette.html">vignette</a>
</li>
<li>
  <a href="../articles/article.html">article</a>
</li>
<li>
  <a href="../news/index.html">news</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rauschenberger/globalSeq">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://bioconductor.org/packages/globalSeq/">
    <span class="fa fa-download"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Omnibus test</h1>
    <small class="dont-index">Source: <a href='https://github.com/rauschenberger/globalSeq/blob/master/R/user_functions.R'><code>R/user_functions.R</code></a></small>
    <div class="hidden name"><code>omnibus.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Test of association between a count response and
one or more covariate sets.
This test may be conceptualised as
a test of overall significance in regression analysis,
where the response variable is overdispersed, and where
the number of explanatory variables (<code>p</code>)
exceeds the sample size (<code>n</code>).
The negative binomial distribution accounts for overdispersion
and a random effect model accounts for high dimensionality
(<code>p</code>&gt;&gt;<code>n</code>).</p>
    
    </div>

    <pre class="usage"><span class='fu'>omnibus</span>(<span class='no'>y</span>, <span class='no'>X</span>, <span class='kw'>offset</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>group</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
        <span class='kw'>mu</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>phi</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
        <span class='kw'>perm</span> <span class='kw'>=</span> <span class='fl'>1000</span>, <span class='kw'>kind</span> <span class='kw'>=</span> <span class='fl'>1</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>y</th>
      <td><p><strong>response variable</strong><strong>:</strong>
numeric vector of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>X</th>
      <td><p><strong>one covariate set</strong><strong>:</strong>
numeric matrix with <code>n</code> rows (samples)
and <code>p</code> columns (covariates);
<br /> <strong>multiple covariate sets</strong><strong>:</strong>
list of numeric matrices with <code>n</code> rows (samples)</p></td>
    </tr>
    <tr>
      <th>offset</th>
      <td><p>numeric vector of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>group</th>
      <td><p>confounding variable<strong>:</strong>
factor of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>mu</th>
      <td><p>mean parameters<strong>:</strong>
numeric vector of length <code>1</code> or <code>n</code></p></td>
    </tr>
    <tr>
      <th>phi</th>
      <td><p>dispersion parameter<strong>:</strong>
non-negative real number</p></td>
    </tr>
    <tr>
      <th>perm</th>
      <td><p>number of iterations<strong>:</strong>
positive integer</p></td>
    </tr>
    <tr>
      <th>kind</th>
      <td><p>computation <strong>:</strong>
number between 0 and 1</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The function returns a dataframe,
with the p-value in the first column,
and the test statistic in the second column.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The user can provide a common <code>mu</code> for all samples
or sample-specific <code>mu</code>, and a common <code>phi</code>.
Setting <code>phi</code> equal to zero is equivalent
to using the Poisson model.
If <code>mu</code> is missing, then <code>mu</code> is estimated from <code>y</code>.
If <code>phi</code> is missing, then <code>mu</code> and <code>phi</code>
are estimated from <code>y</code>.
The <code>offset</code> is only taken into account
for estimating <code>mu</code> or <code>phi</code>.
By default the offset is <code>rep(1,n)</code>.</p>
<p>The user can provide the <strong>confounding</strong> variable <code>group</code>.
Note that each level of <code>group</code> must appear at least twice
in order to allow stratified permutations.</p>
<p>Efficient alternatives to classical <strong>permutation</strong> (<code>kind=1</code>)
are the method of control variates (<code>kind=0</code>)
and permutation in chunks (0 &lt; <code>kind</code> &lt; 1)
<a href='intern.crude.html'>details</a>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>A Rauschenberger, MA Jonker, MA van de Wiel, and RX Menezes (2016).
"Testing for association between RNA-Seq and high-dimensional data",
<em>BMC Bioinformatics</em>. 17:118.
<a href='http://dx.doi.org/10.1186/s12859-016-0961-5'>html</a>
<a href='http://www.biomedcentral.com/content/pdf/s12859-016-0961-5.pdf'>pdf</a>
(open access)</p>
<p>RX Menezes, L Mohammadi, JJ Goeman, and JM Boer (2016).
"Analysing multiple types of molecular profiles simultaneously:
connecting the needles in the haystack",
<em>BMC Bioinformatics</em>. 17:77.
<a href='http://dx.doi.org/10.1186/s12859-016-0926-8'>html</a>
<a href='http://www.biomedcentral.com/content/pdf/s12859-016-0926-8.pdf'>pdf</a>
(open access)</p>
<p>S le Cessie, and HC van Houwelingen (1995).
"Testing the fit of a regression model 
via score tests in random effects models",
<em>Biometrics</em>. 51:600-614.
<a href='http://dx.doi.org/10.2307/2532948'>html</a>
<a href='http://www.jstor.org/stable/pdf/2532948.pdf?acceptTC=true'>pdf</a>
(restricted access)</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>The function <code><a href='proprius.html'>proprius</a></code> calculates
the contributions of individual samples or covariates
to the test statistic.
The function <code><a href='cursus.html'>cursus</a></code> tests for association
between RNA-Seq and local genetic or epigenetic alternations
across the whole genome.
All other functions of the R package <code><a href='globalSeq-package.html'>globalSeq</a></code>
are <code><a href='internal.html'>internal</a></code>.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># simulate high-dimensional data</span>
<span class='no'>n</span> <span class='kw'>&lt;-</span> <span class='fl'>30</span>; <span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fl'>100</span>
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'>rnbinom</span>(<span class='no'>n</span>,<span class='kw'>mu</span><span class='kw'>=</span><span class='fl'>10</span>,<span class='kw'>size</span><span class='kw'>=</span><span class='fl'>1</span>/<span class='fl'>0.25</span>)
<span class='no'>X</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rnorm</span>(<span class='no'>n</span>*<span class='no'>p</span>),<span class='kw'>nrow</span><span class='kw'>=</span><span class='no'>n</span>,<span class='kw'>ncol</span><span class='kw'>=</span><span class='no'>p</span>)

<span class='co'># hypothesis testing</span>
<span class='fu'>omnibus</span>(<span class='no'>y</span>,<span class='no'>X</span>)</div><div class='output co'>#&gt;   pvalue teststat covs
#&gt; 1  0.149 4.170036  100</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://rauschenberger.github.io'>Armin Rauschenberger</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

