<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Decomposition — proprius • globalSeq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/spacelab/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Decomposition — proprius" />

<meta property="og:description" content="Even though the function omnibus tests
a single hypothesis on a whole covariate set,
this function allows to calculate
the individual contributions of n samples or
p covariates to the test statistic." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">globalSeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.11.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">functions</a>
</li>
<li>
  <a href="../articles/vignette.html">vignette</a>
</li>
<li>
  <a href="../articles/article.html">article</a>
</li>
<li>
  <a href="../news/index.html">news</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rauschenberger/globalSeq">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="http://bioconductor.org/packages/globalSeq/">
    <span class="fa fa-download"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Decomposition</h1>
    <small class="dont-index">Source: <a href='https://github.com/rauschenberger/globalSeq/blob/master/R/user_functions.R'><code>R/user_functions.R</code></a></small>
    <div class="hidden name"><code>proprius.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Even though the function <code><a href='omnibus.html'>omnibus</a></code> tests
a single hypothesis on a whole covariate set,
this function allows to calculate
the individual contributions of <code>n</code> samples or
<code>p</code> covariates to the test statistic.</p>
    
    </div>

    <pre class="usage"><span class='fu'>proprius</span>(<span class='no'>y</span>, <span class='no'>X</span>, <span class='no'>type</span>, <span class='kw'>offset</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>group</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
        <span class='kw'>mu</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>phi</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
        <span class='kw'>alpha</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>perm</span> <span class='kw'>=</span> <span class='fl'>1000</span>, <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>y</th>
      <td><p><strong>response variable</strong><strong>:</strong>
numeric vector of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>X</th>
      <td><p><strong>covariate set</strong><strong>:</strong>
numeric matrix with <code>n</code> rows (samples) 
and <code>p</code> columns (covariates)</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>character '<strong>covariates</strong>' or '<strong>samples</strong>'</p></td>
    </tr>
    <tr>
      <th>offset</th>
      <td><p>numeric vector of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>group</th>
      <td><p>confounding variable<strong>:</strong>
factor of length <code>n</code></p></td>
    </tr>
    <tr>
      <th>mu</th>
      <td><p>mean parameters<strong>:</strong>
numeric vector of length <code>1</code> or <code>n</code></p></td>
    </tr>
    <tr>
      <th>phi</th>
      <td><p>dispersion parameter<strong>:</strong>
non-negative real number</p></td>
    </tr>
    <tr>
      <th>alpha</th>
      <td><p>significance level<strong>:</strong> real number between 0 and 1</p></td>
    </tr>
    <tr>
      <th>perm</th>
      <td><p>number of iterations<strong>:</strong>
positive integer</p></td>
    </tr>
    <tr>
      <th>plot</th>
      <td><p>plot of results<strong>:</strong> logical</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>If <code>alpha=NULL</code>, then the function returns a numeric vector,
and else a list of numeric vectors.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The user can provide a common <code>mu</code> for all samples
or sample-specific <code>mu</code>, and a common <code>phi</code>.
Setting <code>phi</code> equal to zero
is equivalent to using the Poisson model.
If <code>mu</code> is missing, then <code>mu</code> is estimated from <code>y</code>.
If <code>phi</code> is missing, then <code>mu</code> and <code>phi</code>
are estimated from <code>y</code>.
The <code>offset</code> is only taken into account
for estimating <code>mu</code> or <code>phi</code>.</p>
<p>The user can provide the confounding variable <code>group</code>.
Note that each level of <code>group</code> must appear at least twice
in order to allow stratified permutations.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>A Rauschenberger, MA Jonker, MA van de Wiel, and RX Menezes (2016).
"Testing for association between RNA-Seq and high-dimensional data",
<em>BMC Bioinformatics</em>. 17:118.
<a href='http://dx.doi.org/10.1186/s12859-016-0961-5'>html</a>
<a href='http://www.biomedcentral.com/content/pdf/s12859-016-0961-5.pdf'>pdf</a>
(open access)</p>
<p>JJ Goeman, SA van de Geer, F de Kort, and HC van Houwelingen (2004).
"A global test for groups of genes:
testing association with a clinical outcome",
<em>Bioinformatics</em>. 20:93-99.
<a href='http://dx.doi.org/10.1093/bioinformatics/btg382'>html</a>
<a href='http://bioinformatics.oxfordjournals.org/content/20/1/93.full.pdf'>pdf</a>
(open access)</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>The function <code><a href='omnibus.html'>omnibus</a></code> tests for associations
between an overdispersed response variable and a high-dimensional
covariate set.
The function <code><a href='cursus.html'>cursus</a></code> tests for association
between RNA-Seq and local genetic or epigenetic alternations
across the whole genome.
All other functions of the R package <code><a href='globalSeq-package.html'>globalSeq</a></code>
are <code><a href='internal.html'>internal</a></code>.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># simulate high-dimensional data</span>
<span class='no'>n</span> <span class='kw'>&lt;-</span> <span class='fl'>30</span>; <span class='no'>p</span> <span class='kw'>&lt;-</span> <span class='fl'>100</span>
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/NegBinomial'>rnbinom</a></span>(<span class='no'>n</span>,<span class='kw'>mu</span><span class='kw'>=</span><span class='fl'>10</span>,<span class='kw'>size</span><span class='kw'>=</span><span class='fl'>1</span>/<span class='fl'>0.25</span>)
<span class='no'>X</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/matrix'>matrix</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>(<span class='no'>n</span>*<span class='no'>p</span>),<span class='kw'>nrow</span><span class='kw'>=</span><span class='no'>n</span>,<span class='kw'>ncol</span><span class='kw'>=</span><span class='no'>p</span>)

<span class='co'># decomposition</span>
<span class='fu'>proprius</span>(<span class='no'>y</span>,<span class='no'>X</span>,<span class='kw'>type</span><span class='kw'>=</span><span class='st'>"samples"</span>)</div><div class='img'><img src='proprius-1.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt;  [1] -1.473829651 -1.621989536  0.389933792 -1.083847301  1.892962633
#&gt;  [6] -0.960717613  1.721503713 -0.122347060  2.878808190 -1.212848450
#&gt; [11]  0.265357110 -1.419499667 -0.945988913 -1.218230787 -1.565611811
#&gt; [16]  1.924302618 -0.749555757 -1.266216291 -0.778061491 -0.009607322
#&gt; [21] -1.164619359 -1.077795751 -1.272294761 -1.232923857 -1.496262593
#&gt; [26] -0.515822170  2.721173151 -1.648988791  0.638509691 -1.424787341</div><div class='input'><span class='fu'>proprius</span>(<span class='no'>y</span>,<span class='no'>X</span>,<span class='kw'>type</span><span class='kw'>=</span><span class='st'>"covariates"</span>)</div><div class='img'><img src='proprius-2.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt;   [1] -0.061952584  0.592890632 -0.092819492  0.084168673 -0.397585279
#&gt;   [6] -0.183539882  0.031889996 -0.426508046 -0.259698391 -0.265361798
#&gt;  [11] -0.317232615  0.072810328 -0.239941836 -0.245169345 -0.365328320
#&gt;  [16] -0.167861221  0.019013502 -0.293306936 -0.074805315 -0.269415642
#&gt;  [21]  0.270934190 -0.086943070 -0.619497569  0.755689349 -0.227954044
#&gt;  [26] -0.205152081 -0.223616292 -0.376515968 -0.197909498  0.272830697
#&gt;  [31] -0.417295924 -0.572140052 -0.139501582 -0.418759569 -0.195663061
#&gt;  [36] -0.188234195 -0.521401610 -0.415634949 -0.346467928 -0.349298405
#&gt;  [41] -0.270644797 -0.058625362 -0.435316923 -0.261076519 -0.314005429
#&gt;  [46] -0.376683458 -0.544577063  0.139635838 -0.452980287  0.660662887
#&gt;  [51] -0.284095283 -0.529192814 -0.293848069 -0.152128001 -0.451823530
#&gt;  [56]  0.598375469  0.024277366  0.386251570  0.118416105 -0.275617185
#&gt;  [61] -0.243857126 -0.235047343  0.008511704 -0.275980122 -0.173835916
#&gt;  [66] -0.270343598 -0.433204607 -0.301079993 -0.452220868 -0.437245563
#&gt;  [71]  0.909077707  1.224714475  0.035826449 -0.203264859  1.192802248
#&gt;  [76]  0.094335123 -0.245741028  0.620533156 -0.302885587 -0.508894499
#&gt;  [81] -0.331829320 -0.218798231 -0.360255918 -0.162935165 -0.323970670
#&gt;  [86] -0.143899832 -0.247114748 -0.161513535 -0.044980916 -0.136249689
#&gt;  [91]  0.648363088  0.604569515  0.332039489 -0.302868514 -0.468019955
#&gt;  [96] -0.131218846 -0.247416189 -0.281275540 -0.009120144 -0.009719390</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://rauschenberger.github.io'>Armin Rauschenberger</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

